### PR TITLE
Update Sortable to support tablemenu plugin

### PR DESCRIPTION
**Pull request by JulianPage**
P2X-1227 Update sortable to support tablemenu plugin

### CODE REVIEW

File: Sortable.js [hunk 1]
```diff
@@ -417,37 +460,54 @@
 					// Bind the events: dragstart/dragend
 					_this._triggerDragStart(evt, touch);
 
-					// Drag start event
-					_dispatchEvent(_this, rootEl, 'choose', dragEl, rootEl, oldIndex);
+                    // Drag start event
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 2]
```diff
@@ -526,12 +618,22 @@
 
 				if (!supportCssPointerEvents) {
 					_css(ghostEl, 'display', 'none');
-				}
-
+                }
+
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 3]
```diff
@@ -576,21 +688,39 @@
...
 
-				_css(ghostEl, 'webkitTransform', translate3d);
-				_css(ghostEl, 'mozTransform', translate3d);
-				_css(ghostEl, 'msTransform', translate3d);
-				_css(ghostEl, 'transform', translate3d);
+				/**
+				 * @start_change
+				 * Set touchEvt _dragRow if tapEvt has it
+				 */
+                if ( tapEvt.hasOwnProperty( '_dragRow' ) === true ) {
```
Comment (GMiLL5) [inline Sortable.js]: Spacing down here

File: Sortable.js [hunk 4]
```diff
@@ -576,21 +688,39 @@
...
+                    touchEvt._dragRow = tapEvt._dragRow;
+				}
+				/** @end_change */
+
+                _css(ghostEl, 'webkitTransform', translate3d);
```
Comment (GMiLL5) [inline Sortable.js]: Why did we move this?

File: Sortable.js [hunk 5]
```diff
@@ -601,11 +731,31 @@
 				var rect = dragEl.getBoundingClientRect(),
 					css = _css(dragEl),
 					options = this.options,
-					ghostRect;
-
-				ghostEl = dragEl.cloneNode(true);
-
-				_toggleClass(ghostEl, options.ghostClass, false);
+                    ghostRect;
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 6]
```diff
@@ -601,11 +731,31 @@
...
+
+                ghostEl = dragEl.cloneNode(true);
```
Comment (GMiLL5) [inline Sortable.js]: spacing

File: Sortable.js [hunk 7]
```diff
@@ -601,11 +731,31 @@
...
+                
+				/**
+				 * @start_change
+				 * Call _toggleClass on ghostEl, turning on chosenClass
+				 */
+				if ( options.tableMode === true ) {
+					_toggleClass( ghostEl, options.chosenClass, false );
```
Comment (GMiLL5) [inline Sortable.js]: Is chosenClass a new option? Are we not able  to use ghostClass?

File: Sortable.js [hunk 8]
```diff
@@ -629,9 +779,10 @@
 				options.fallbackOnBody && document.body.appendChild(ghostEl) || rootEl.appendChild(ghostEl);
 
 				// Fixing dimensions.
-				ghostRect = ghostEl.getBoundingClientRect();
+                ghostRect = ghostEl.getBoundingClientRect();
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 9]
```diff
@@ -629,9 +779,10 @@
...
+                
 				_css(ghostEl, 'width', rect.width * 2 - ghostRect.width);
-				_css(ghostEl, 'height', rect.height * 2 - ghostRect.height);
+                _css(ghostEl, 'height', rect.height * 2 - ghostRect.height);
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 10]
```diff
@@ -652,23 +811,32 @@
 
 				rootEl.insertBefore(cloneEl, dragEl);
 				_dispatchEvent(this, rootEl, 'clone', dragEl);
-			}
-
-			_toggleClass(dragEl, options.dragClass, true);
+            }
+            
+            _toggleClass(dragEl, options.dragClass, true);
```
Comment (GMiLL5) [inline Sortable.js]: Spacing on this & two lines above

File: Sortable.js [hunk 11]
```diff
@@ -706,7 +874,7 @@
 			}
 
 			moved = true;
-
+
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 12]
```diff
@@ -717,27 +885,28 @@
 							group.checkPut(this, activeSortable, dragEl, evt)
 						)
 					)
-				) &&
-				(evt.rootEl === void 0 || evt.rootEl === this.el) // touch fallback
+                ) &&
```
Comment (GMiLL5) [inline Sortable.js]: spacing

File: Sortable.js [hunk 13]
```diff
@@ -717,27 +885,28 @@
...
+				(evt.rootEl === void 0 || evt.rootEl === this.el ) // touch fallback
 			) {
-				// Smart auto-scrolling
+                // Smart auto-scrolling
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 14]
```diff
@@ -717,27 +885,28 @@
...
 				_autoScroll(evt, options, this.el);
 
 				if (_silent) {
 					return;
 				}
 
-				target = _closest(evt.target, options.draggable, el);
+                target = _closest(evt.target, options.draggable, el);
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 15]
```diff
@@ -717,27 +885,28 @@
...
+                
 				dragRect = dragEl.getBoundingClientRect();
-
+
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 16]
```diff
@@ -717,27 +885,28 @@
...
+                
 				if (putSortable !== this) {
 					putSortable = this;
 					isMovingBetweenSortable = true;
 				}
 
 				if (revert) {
 					_cloneHide(activeSortable, true);
-					parentEl = rootEl; // actualization
+                    parentEl = rootEl; // actualization
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 17]
```diff
@@ -749,7 +918,6 @@
 					return;
 				}
 
-
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 18]
```diff
@@ -777,9 +945,18 @@
...
-				else if (target && !target.animated && target !== dragEl && (target.parentNode[expando] !== void 0)) {
-					if (lastEl !== target) {
+                }
+				/**
+				 * @start_change
+				 * When in tableMode, alter the logic slightly
+				 * Original:
+				 *  else if (target && !target.animated && target !== dragEl && (target.parentNode[expando] !== void 0)) {
+				 */
+                else if (target && !target.animated && ( ( this.options.tableMode === false && target !== dragEl ) || this.options.tableMode === true ) && (target.parentNode[expando] !== void 0)) {
```
Comment (GMiLL5) [inline Sortable.js]: It looks like this is all outside of the block it  used to be - is that going to cause problems?

File: Sortable.js [hunk 19]
```diff
@@ -814,8 +991,8 @@
 						after = (nextSibling !== dragEl) && !isLong || halfway && isLong;
 					}
 
-					var moveVector = _onMove(rootEl, el, dragEl, dragRect, target, targetRect, evt, after);
-
+                    var moveVector = _onMove(rootEl, el, dragEl, dragRect, target, targetRect, evt, after);
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 20]
```diff
@@ -814,8 +991,8 @@
...
+
```
Comment (GMiLL5) [inline Sortable.js]: spacing

File: Sortable.js [hunk 21]
```diff
@@ -824,27 +1001,46 @@
...
 						setTimeout(_unsilent, 30);
 
-						_cloneHide(activeSortable, isOwner);
-
-						if (!dragEl.contains(el)) {
-							if (after && !nextSibling) {
-								el.appendChild(dragEl);
-							} else {
-								target.parentNode.insertBefore(dragEl, after ? nextSibling : target);
+                        _cloneHide(activeSortable, isOwner);
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 22]
```diff
@@ -824,27 +1001,46 @@
...
 
 						parentEl = dragEl.parentNode; // actualization
 
-						this._animate(dragRect, dragEl);
-						this._animate(targetRect, target);
+						/**
+						 * @start_change
+						 * Add this.options.tableMode === false if block
+						 */
+						if ( this.options.tableMode === false ) {
```
Comment (GMiLL5) [inline Sortable.js]: Is this so  we don't animate the dragging? Is there an option to disable this instead?

File: Sortable.js [hunk 23]
```diff
@@ -824,27 +1001,46 @@
...
+						 * @end_change
+						 */
 					}
 				}
 			}
 		},
 
 		_animate: function (prevRect, target) {
-			var ms = this.options.animation;
+            var ms = this.options.animation;
```
Comment (GMiLL5) [inline Sortable.js]: spacing

File: Sortable.js [hunk 24]
```diff
@@ -857,8 +1053,8 @@
 				_css(target, 'transform', 'translate3d('
 					+ (prevRect.left - currentRect.left) + 'px,'
 					+ (prevRect.top - currentRect.top) + 'px,0)'
-				);
-
+                );
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 25]
```diff
@@ -919,16 +1142,36 @@
...
+						_off(dragEl, 'dragend', this, options);
 					}
 
-					_disableDraggable(dragEl);
-					dragEl.style['will-change'] = '';
-
-					// Remove class's
-					_toggleClass(dragEl, this.options.ghostClass, false);
-					_toggleClass(dragEl, this.options.chosenClass, false);
+                    _disableDraggable(dragEl);
```
Comment (GMiLL5) [inline Sortable.js]: Spacing in this block (see down)

File: Sortable.js [hunk 26]
```diff
@@ -919,16 +1142,36 @@
...
+					 * @start_change
+					 * When in tableMode, do not use ghostClass on dragEl
+					 * Original:
+					 *  _toggleClass(dragEl, this.options.ghostClass, false);	
+					 */
+					if ( options.tableMode === false ) {
+						_toggleClass(dragEl, this.options.ghostClass, false);	
+					}
 
+                    _toggleClass(dragEl, this.options.chosenClass, false);
```
Comment (GMiLL5) [inline Sortable.js]: Until here (Spacing note above)

File: Sortable.js [hunk 27]
```diff
@@ -1210,18 +1462,75 @@
...
+		/** 
+		 * @start_change
+		 * When in tableMode, use a capture: true captureMode
+		 * Original:
+		 *  el.addEventListener(event, fn, captureMode);
+		 */
+		var mode = captureMode;
+
+		if ( options && options.hasOwnProperty( 'tableMode' ) === true && options.tableMode === true ) {
+			mode = { capture: true, passive: false };
```
Comment (GMiLL5) [inline Sortable.js]: Couldn't we just  add a 'captureMode' param to the plugin?

File: Sortable.js [hunk 28]
```diff
@@ -1210,18 +1462,75 @@
...
+		/** 
+		 * @start_change
+		 * When in tableMode, use a capture: true captureMode
+		 * Original:
+		 *  el.removeEventListener(event, fn, captureMode);
+		 */
+		var mode = captureMode;
+
+		if ( options && options.hasOwnProperty( 'tableMode' ) === true && options.tableMode === true ) {
+			mode = { capture: true, passive: false };
```
Comment (GMiLL5) [inline Sortable.js]: Couldn't  we just add a captureMode param to the plugin?

File: Sortable.js [hunk 29]
```diff
@@ -1210,18 +1462,75 @@
...
 
 	function _toggleClass(el, name, state) {
-		if (el) {
+        /**
+		 * @start_change
+		 * Check for both el and name, not just el
+		 * Original:
+		 *  if (el) {}
+		 */
+		if (el && name) {
```
Comment (GMiLL5) [inline Sortable.js]: Why do we need to check for `name` here now too?

File: Sortable.js [hunk 30]
```diff
@@ -1291,7 +1607,18 @@
 		evt.clone = cloneEl;
 
 		evt.oldIndex = startIndex;
-		evt.newIndex = newIndex;
+        evt.newIndex = newIndex;
```
Comment (GMiLL5) [inline Sortable.js]: Spacing (until the end of the block)

File: Sortable.js [hunk 31]
```diff
@@ -498,18 +567,41 @@
...
+		 * @end_change 
+		 */
 			if (rootEl && dragEl) {
 				var options = this.options;
 
 				// Apply effect
-				_toggleClass(dragEl, options.ghostClass, true);
+				/**
+				 * @start_change
+				 * Call _toggleClass on options.dragClass, not options.ghostClass
```
[the remark this answers is not in the document]
Comment (JulianPage) [reply Sortable.js]: This was (as far as i can understand it) a bug;  the original code added the options.ghostClass class to dragEl, which doesn't make any sense to me, and also didn't work out with us using a ghostClass of 'hidden'! I will check the logic on why i swapped the 3rd param around

File: Sortable.js [hunk 32]
```diff
@@ -576,21 +688,39 @@
...
-
-					this._dragStarted();
+                    }
+                    /**
+					 * @start_change
+					 * Include evt parameter
+					 * Original:
+					 *  this._dragStarted();
+					 */
+					this._dragStarted(evt);
```
[the remark this answers is not in the document]
Comment (JulianPage) [reply Sortable.js]: Either spacing or the positioning of my @start_change notes i think, the call to _dragStarted is in the same block that it used to be, but has an additional parameter (evt) now

File: Sortable.js (hunk 22 above)
Comment (JulianPage) [reply Sortable.js]: taken this section out, unnecessary; we're setting animation to 0 anyway, which causes an immediate exit from the _animate function.

File: Sortable.js (hunk 27 above)
Comment (JulianPage) [reply Sortable.js]: done!

File: Sortable.js (hunk 28 above)
Comment (JulianPage) [reply Sortable.js]: done also

File: Sortable.js (hunk 31 above)
[the remark this answers is not in the document]
Comment (GMiLL5) [reply Sortable.js]: We'll just need to make sure that this isn't  breaking any of  the other  sortable instances (See section-clone, section-list-ghost, placeholder)

File: Sortable.js [hunk 33]
```diff
@@ -526,12 +618,22 @@
...
+                
 				var target = document.elementFromPoint(touchEvt.clientX, touchEvt.clientY),
 					parent = target,
 					i = touchDragOverListeners.length;
 
+				/**
+				 * @start_change
+				 * If the event originated from _dragRow, use it for parent/target
+				 */
+				if ( touchEvt.hasOwnProperty( '_dragRow' ) === true ) {
```
[the remark this answers is not in the document]
Comment (GMiLL5) [reply Sortable.js]: @JulianPage

File: Sortable.js [hunk 34]
```diff
@@ -587,6 +691,15 @@
 				moved = true;
 				touchEvt = touch;
 
+				/**
+				 * @start_change
+				 * Set touchEvt _dragRow if tapEvt has it
+				 */
+				if ( tapEvt.hasOwnProperty( '_dragRow' ) === true ) {
```
Comment (GMiLL5) [inline Sortable.js]: Same idea as above - can we modify this to be a passed param, a function, etc?

File: Sortable.js [hunk 35]
```diff
@@ -826,13 +955,23 @@
...
+						 * @start_change
+						 * Add if ( this.options.draggable !== 'tr' ) block
+						 */
+						if ( this.options.draggable !== 'tr' ) {
+							if (!dragEl.contains(el)) {
+								if (after && !nextSibling) {
+									el.appendChild(dragEl);
+								} else {
+									target.parentNode.insertBefore(dragEl, after ? nextSibling : target);
+									console.log('added ', dragEl, ' to target.parentNode');
```
Comment (GMiLL5) [inline Sortable.js]: Remove logging

File: Sortable.js [hunk 36]
```diff
@@ -1209,19 +1367,62 @@
...
+		var mode = captureMode;
+
+		if ( opts && opts.captureMode === true ) {
+			mode = {
+				capture: true,
+				passive: false,
+			};
+		}
+
+		el && el.addEventListener(event, fn, mode);
```
Comment (GMiLL5) [inline Sortable.js]: Is it possible for el not to exist?

File: Sortable.js [hunk 37]
```diff
@@ -1209,19 +1367,62 @@
...
+		var mode = captureMode;
 
+		if ( opts && opts.captureMode === true ) {
+			mode = {
+				capture: true,
+				passive: false,
+			};
+		}
+
+		el && el.removeEventListener(event, fn, mode);
```
Comment (GMiLL5) [inline Sortable.js]: Is it possible for el not to exist?

File: Sortable.js [hunk 38]
```diff
@@ -1209,19 +1367,62 @@
...
 
 	function _toggleClass(el, name, state) {
-		if (el) {
+		/**
+		 * @start_change
+		 * Check for both el and name, not just el
+		 * Original:
+		 if (el) {}
+		 */
+		if (el && name) {
```
Comment (GMiLL5) [inline Sortable.js]: Did we find out why we need to check for name?

File: Sortable.js [hunk 39]
```diff
@@ -1317,6 +1536,14 @@
 		evt.related = targetEl || toEl;
 		evt.relatedRect = targetRect || toEl.getBoundingClientRect();
 		evt.willInsertAfter = willInsertAfter;
+		/** @start_change
+		 * Include original event click coordinates (clientY and clientX)
+		 */
+		evt.clientY = originalEvt.clientY;
```
Comment (GMiLL5) [inline Sortable.js]: Not sure where  to  leave this - but don't we need to set clientX to 0 if we disable horizontal dragging? (should add  the  same for vertical)

File: Sortable.js [hunk 40]
```diff
@@ -267,7 +274,21 @@
...
+			 * Add trailing comma after fallbackOffset value
+			 * Add captureMode (default false)
+			 * Add dragWillChange (default true)
+			 * Add allowDragX and allowDragY (both default to true)
+			 */
+			fallbackOffset: {x: 0, y: 0},
+			captureMode: false,
+			dragWillChange: true,
+			allowDragX: true,
+			allowDragY: true
```
Comment (GMiLL5) [inline Sortable.js]: Set an `elementFromPoint` in the defaults array and set it to false

File: Sortable.js [hunk 41]
```diff
@@ -527,10 +609,26 @@
...
+				 * Add support for elementFromPoint callback function
+				 * Original:
 				var target = document.elementFromPoint(touchEvt.clientX, touchEvt.clientY),
 					parent = target,
 					i = touchDragOverListeners.length;
+				 */
+				var i = touchDragOverListeners.length,
+					target, parent;
+
+                if (this.options.hasOwnProperty('elementFromPoint') === true && typeof this.options.elementFromPoint === 'function') {
```
Comment (GMiLL5) [inline Sortable.js]: Add this to the  defaults array (See note above) & just check if `typeof this.options.elementFromPoint === 'function'`

File: Sortable.js [hunk 42]
```diff
@@ -603,7 +720,21 @@
 					options = this.options,
 					ghostRect;
 
+				/**
+				 * @start_change
+				 * Use 'ghost' options callback to construct ghost (if provided)
+				 * Original:
 				ghostEl = dragEl.cloneNode(true);
+				 */
+				if ( options.ghost && typeof options.ghost === 'function' ) {
```
Comment (GMiLL5) [inline Sortable.js]: Same as the elementFromPoint,  add to the options array and default to false

File: Sortable.js [hunk 43]
```diff
@@ -527,10 +615,26 @@
...
+				 * Add support for elementFromPoint callback function
+				 * Original:
 				var target = document.elementFromPoint(touchEvt.clientX, touchEvt.clientY),
 					parent = target,
 					i = touchDragOverListeners.length;
+				 */
+				var i = touchDragOverListeners.length,
+					target, parent;
+
+                if (typeof this.options.elementFromPoint === 'function') {
```
Comment (GMiLL5) [inline Sortable.js]: Spacing

File: Sortable.js [hunk 44]
```diff
@@ -587,6 +710,34 @@
...
 				touchEvt = touch;
 
+				/**
+				 * @start_change
+				 * Constrain if data-constrain attribute is present on ghostEl
+				 * Also fire onGhostMove if configured in options to do so
+				 */
+				var d_rect, root_rect;
+
+				if (ghostEl.hasAttribute('data-constrain') === true && dy !== 0) {
```
Comment (GMiLL5) [inline Sortable.js]: Is this something we could do with a param/callback option? Seems forced in to support something we're doing

**Pull request merged**